### PR TITLE
Fix JSON/Dynamic Native serialization with old server and new client and add a setting to fallback to V1 version 

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
After https://github.com/ClickHouse/ClickHouse/pull/70442 (not included in any release yet) we use new Native serialization for JSON and Dynamic, but it breaks new client - old server communication. This PR fixes it.
Also add a setting `merge_tree_use_v1_object_and_dynamic_serialization` that allows to use V1 serialization in newer versions (it can be used during upgrades to new version to be able to rollback).

Closes https://github.com/ClickHouse/ClickHouse/issues/71819

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 69e4f93a2adb1b8a88e6239406f6a4f122c08010 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71816/69e4f93a2adb1b8a88e6239406f6a4f122c08010/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71779/621cb60446cb17f0366f49b86c3432eed5db3716/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Formats/NativeWriter.cpp [hunk 1]
```diff
@@ -70,6 +70,7 @@ static void writeData(const ISerialization & serialization, const ColumnPtr & co
     settings.low_cardinality_max_dictionary_size = 0;
     settings.data_types_binary_encoding = format_settings && format_settings->native.encode_types_in_binary_format;
     settings.write_json_as_string = format_settings && format_settings->native.write_json_as_string;
+    settings.use_v1_object_and_dynamic_serialization = client_revision < DBMS_MIN_REVISION_WITH_V2_DYNAMIC_AND_JSON_SERIALIZATION;
```
Comment (kssenii) [inline src/Formats/NativeWriter.cpp:73]: so it's rather old client - new server fix, not new client - old server (like PR description says)?

Comment (Avogar) [reply src/Formats/NativeWriter.cpp:73]: No, it's actually when new client sends data to old server using Native protocol. In this case `client_revision` here will be the revision of the server we send the data to.

**Review by kssenii — APPROVED** [review]

File: src/Core/SettingsChangesHistory.cpp [hunk 2]
```diff
@@ -77,6 +77,7 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
             {"backup_restore_keeper_max_retries_while_handling_error", 0, 20, "New setting."},
             {"backup_restore_finish_timeout_after_error_sec", 0, 180, "New setting."},
             {"parallel_replicas_local_plan", false, true, "Use local plan for local replica in a query with parallel replicas"},
+            {"merge_tree_use_v1_object_and_dynamic_serialization", true, false, "Add new serialization V2 version for JSON and Dynamic types"}
```
Comment (Algunenano) [inline src/Core/SettingsChangesHistory.cpp:80]: Doesn't it make more sense to make it a server setting? I'm not sure we want this to be available for changing during queries, for example.

Comment (Avogar) [reply src/Core/SettingsChangesHistory.cpp:80]: Does `compatibility` setting allows to change server settings? If not, then better to have it here. Also from the code, this setting takes affect only when enabled for global context (probably need to make a note that changing this settings requires server restart)

**Pull request merged**